File: backend-java/src/main/java/com/simplemdm/service/ApprovalService.java
package com.simplemdm.service;

import com.simplemdm.dto.PersonnelDTO;
import com.simplemdm.model.*;
import com.simplemdm.repository.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
public class ApprovalService {

    private final WfApprovalRepository approvalRepo;
    private final MdmPersonnelRepository personnelRepo;
    private final SysUserRepository userRepo;
    private final SysApproverDeptRepository approverDeptRepo;
    private final PersonnelService personnelService;
    private final PushService pushService;
    private final ObjectMapper mapper = new ObjectMapper();

    public ApprovalService(WfApprovalRepository approvalRepo, MdmPersonnelRepository personnelRepo,
                           SysUserRepository userRepo, SysApproverDeptRepository approverDeptRepo,
                           PersonnelService personnelService, PushService pushService) {
        this.approvalRepo = approvalRepo;
        this.personnelRepo = personnelRepo;
        this.userRepo = userRepo;
        this.approverDeptRepo = approverDeptRepo;
        this.personnelService = personnelService;
        this.pushService = pushService;
    }

    @Transactional
    public WfApproval createApprovalForCreate(Long submitterId, PersonnelDTO dto) {
        MdmPersonnel p = personnelService.createFromApproval(dto);

        Map<String, Map<String, Object>> changeData = new HashMap<>();
        for (java.lang.reflect.Field f : PersonnelDTO.class.getDeclaredFields()) {
            try {
                f.setAccessible(true);
                Object val = f.get(dto);
                if (val != null && !"id".equals(f.getName()) && !"version".equals(f.getName()) && !"status".equals(f.getName())) {
                    changeData.put(f.getName(), Map.of("old", null, "new", val));
                }
            } catch (Exception ignored) {}
        }

        Long approverId = findApproverForDepartment(p.getDepartment());

        WfApproval approval = new WfApproval();
        approval.setPersonnelId(p.getId());
        approval.setWorkflowType("create");
        approval.setSubmitterId(submitterId);
        approval.setApproverId(approverId);
        approval.setStatus("pending");
        try {
            approval.setChangeData(mapper.writeValueAsString(changeData));
        } catch (Exception e) { throw new RuntimeException(e); }
        return approvalRepo.save(approval);
    }

    @Transactional
    public WfApproval createApprovalForUpdate(Long personnelId, Long submitterId, PersonnelDTO dto) {
        MdmPersonnel p = personnelRepo.findById(personnelId).orElse(null);
        if (p == null || "pending_approval".equals(p.getStatus())) return null;

        Map<String, Object> diffResult = personnelService.computeDiff(p, dto);
        if (diffResult == null) return null;

        @SuppressWarnings("unchecked")
        Map<String, Map<String, Object>> diff = (Map<String, Map<String, Object>>) diffResult.get("diff");
        if (diff.isEmpty()) return null;

        p.setStatus("pending_approval");
        personnelRepo.save(p);

        Long approverId = findApproverForDepartment(p.getDepartment());

        WfApproval approval = new WfApproval();
        approval.setPersonnelId(personnelId);
        approval.setWorkflowType("update");
        approval.setSubmitterId(submitterId);
        approval.setApproverId(approverId);
        approval.setStatus("pending");
        try {
            approval.setChangeData(mapper.writeValueAsString(diff));
        } catch (Exception e) { throw new RuntimeException(e); }
        return approvalRepo.save(approval);
    }

    @Transactional(readOnly = true)
    public Page<Map<String, Object>> listApprovals(Long userId, String listType, String statusFilter,
                                                    int page, int pageSize, List<String> approverDepts) {
        Pageable pageable = PageRequest.of(page - 1, pageSize, Sort.by(Sort.Direction.DESC, "id"));
        Page<WfApproval> approvals;

        if ("pending_my".equals(listType) && approverDepts != null) {
            // Find all approvers in the user's managed departments
            List<Long> approverIds = new ArrayList<>();
            approverIds.add(userId);
            approvals = approvalRepo.findByApproverIdInAndStatus(approverIds, "pending", pageable);
        } else if ("pending_my".equals(listType)) {
            approvals = approvalRepo.findByApproverIdAndStatus(userId, "pending", pageable);
        } else if ("my_submitted".equals(listType)) {
            approvals = approvalRepo.findBySubmitterId(userId, pageable);
        } else {
            approvals = approvalRepo.findAll(pageable);
        }

        return approvals.map(this::enrichApproval);
    }

    private Map<String, Object> enrichApproval(WfApproval a) {
        Map<String, Object> result = new HashMap<>();
        result.put("id", a.getId());
        result.put("personnel_id", a.getPersonnelId());
        result.put("personnel_name", personnelRepo.findById(a.getPersonnelId()).map(MdmPersonnel::getName).orElse(""));
        result.put("workflow_type", a.getWorkflowType());
        result.put("submitter_id", a.getSubmitterId());
        result.put("submitter_name", userRepo.findById(a.getSubmitterId()).map(SysUser::getRealName).orElse(""));
        result.put("approver_id", a.getApproverId());
        result.put("approver_name", a.getApproverId() != null ? userRepo.findById(a.getApproverId()).map(SysUser::getRealName).orElse("") : "");
        result.put("status", a.getStatus());
        result.put("change_data", a.getChangeData());
        result.put("submit_time", a.getSubmitTime() != null ? a.getSubmitTime().toString() : null);
        result.put("approve_time", a.getApproveTime() != null ? a.getApproveTime().toString() : null);
        result.put("approve_comment", a.getApproveComment());
        result.put("withdrawn_time", a.getWithdrawnTime() != null ? a.getWithdrawnTime().toString() : null);
        result.put("created_at", a.getCreatedAt() != null ? a.getCreatedAt().toString() : null);
        return result;
    }

    @Transactional
    public WfApproval approve(Long approvalId, String comment) {
        WfApproval a = approvalRepo.findById(approvalId).orElse(null);
        if (a == null || !"pending".equals(a.getStatus())) return null;

        a.setStatus("approved");
        a.setApproveTime(LocalDateTime.now());
        a.setApproveComment(comment);

        MdmPersonnel p = personnelRepo.findById(a.getPersonnelId()).orElse(null);
        if (p != null && a.getChangeData() != null) {
            personnelService.applyChanges(p, a.getChangeData());
        }

        approvalRepo.save(a);

        // Trigger push
        pushService.executePush(a);

        return a;
    }

    @Transactional
    public WfApproval reject(Long approvalId, String comment) {
        WfApproval a = approvalRepo.findById(approvalId).orElse(null);
        if (a == null || !"pending".equals(a.getStatus())) return null;

        a.setStatus("rejected");
        a.setApproveTime(LocalDateTime.now());
        a.setApproveComment(comment);

        MdmPersonnel p = personnelRepo.findById(a.getPersonnelId()).orElse(null);
        if (p != null) {
            if ("create".equals(a.getWorkflowType())) {
                p.setStatus("inactive");
            } else {
                p.setStatus("active");
            }
            personnelRepo.save(p);
        }

        return approvalRepo.save(a);
    }

    @Transactional
    public WfApproval withdraw(Long approvalId, Long userId) {
        WfApproval a = approvalRepo.findById(approvalId).orElse(null);
        if (a == null || !"pending".equals(a.getStatus()) || !a.getSubmitterId().equals(userId)) return null;

        a.setStatus("withdrawn");
        a.setWithdrawnTime(LocalDateTime.now());

        MdmPersonnel p = personnelRepo.findById(a.getPersonnelId()).orElse(null);
        if (p != null) {
            if ("create".equals(a.getWorkflowType())) {
                p.setStatus("inactive");
            } else {
                p.setStatus("active");
            }
            personnelRepo.save(p);
        }

        return approvalRepo.save(a);
    }

    public Map<String, Object> getApprovalDetail(Long id) {
        WfApproval a = approvalRepo.findById(id).orElse(null);
        if (a == null) return null;
        return enrichApproval(a);
    }

    private Long findApproverForDepartment(String department) {
        // Find approver assigned to this department
        List<SysApproverDept> assignments = approverDeptRepo.findByDepartment(department);
        if (!assignments.isEmpty()) {
            return assignments.get(0).getUserId();
        }
        // Fallback: first active admin
        List<SysUser> users = userRepo.findAll();
        for (SysUser u : users) {
            if ("active".equals(u.getStatus()) && Boolean.TRUE.equals(u.getIsAdmin())) {
                return u.getId();
            }
        }
        return null;
    }
}

File: backend/app/services/approval_service.py
"""Approval workflow state machine."""

import json
from datetime import datetime, timezone

from sqlalchemy import select, func
from sqlalchemy.orm import Session

from app.models.approval import WfApproval
from app.models.personnel import MdmPersonnel
from app.models.user import SysUser
from app.schemas.personnel import PersonnelCreate, PersonnelUpdate
from app.services import personnel_service


def create_approval_for_create(
    db: Session,
    submitter_id: int,
    personnel_data: PersonnelCreate,
) -> WfApproval:
    """Create a personnel record as pending_approval, and create a 'create' approval."""
    p = MdmPersonnel(**personnel_data.model_dump(), status="pending_approval")
    db.add(p)
    db.flush()  # get p.id

    change_data = {}
    for field, value in personnel_data.model_dump().items():
        change_data[field] = {"old": None, "new": value}

    approval = WfApproval(
        personnel_id=p.id,
        workflow_type="create",
        submitter_id=submitter_id,
        approver_id=_find_approver(db),
        status="pending",
        change_data=json.dumps(change_data, ensure_ascii=False),
    )
    db.add(approval)
    db.commit()
    db.refresh(approval)
    return approval


def create_approval_for_update(
    db: Session,
    personnel_id: int,
    submitter_id: int,
    data: PersonnelUpdate,
) -> WfApproval | None:
    """Compute diff, mark personnel as pending_approval, create approval. Returns None if no changes."""
    personnel = db.get(MdmPersonnel, personnel_id)
    if not personnel or personnel.status == "pending_approval":
        return None

    diff = personnel_service.compute_diff(personnel, data)
    if diff is None:
        return None

    personnel.status = "pending_approval"

    approval = WfApproval(
        personnel_id=personnel_id,
        workflow_type="update",
        submitter_id=submitter_id,
        approver_id=_find_approver(db),
        status="pending",
        change_data=json.dumps(diff, ensure_ascii=False),
    )
    db.add(approval)
    db.commit()
    db.refresh(approval)
    return approval


def list_approvals(
    db: Session,
    current_user_id: int,
    list_type: str = "all",
    status_filter: str = "",
    page: int = 1,
    page_size: int = 10,
) -> tuple[list[dict], int]:
    """Return approvals with joined user/personnel names."""
    q = select(WfApproval)

    if list_type == "pending_my":
        q = q.where(WfApproval.approver_id == current_user_id, WfApproval.status == "pending")
    elif list_type == "my_submitted":
        q = q.where(WfApproval.submitter_id == current_user_id)
    # else: "all" — no filter

    if status_filter:
        q = q.where(WfApproval.status == status_filter)

    # Count
    count_q = select(func.count()).select_from(q.subquery())
    total = db.execute(count_q).scalar() or 0

    # Paginate
    q = q.order_by(WfApproval.id.desc()).offset((page - 1) * page_size).limit(page_size)
    approvals = list(db.execute(q).scalars().all())

    # Enrich with names
    results = []
    for a in approvals:
        submitter = db.get(SysUser, a.submitter_id)
        approver = db.get(SysUser, a.approver_id) if a.approver_id else None
        personnel = db.get(MdmPersonnel, a.personnel_id)
        results.append({
            "id": a.id,
            "personnel_id": a.personnel_id,
            "personnel_name": personnel.name if personnel else "",
            "workflow_type": a.workflow_type,
            "submitter_id": a.submitter_id,
            "submitter_name": submitter.real_name if submitter else "",
            "approver_id": a.approver_id,
            "approver_name": approver.real_name if approver else "",
            "status": a.status,
            "change_data": a.change_data,
            "submit_time": a.submit_time,
            "approve_time": a.approve_time,
            "approve_comment": a.approve_comment,
            "withdrawn_time": a.withdrawn_time,
            "created_at": a.created_at,
        })
    return results, total


def get_approval_detail(db: Session, approval_id: int) -> dict | None:
    """Get a single approval with all joined data."""
    a = db.get(WfApproval, approval_id)
    if not a:
        return None
    submitter = db.get(SysUser, a.submitter_id)
    approver = db.get(SysUser, a.approver_id) if a.approver_id else None
    personnel = db.get(MdmPersonnel, a.personnel_id)
    return {
        "id": a.id,
        "personnel_id": a.personnel_id,
        "personnel_name": personnel.name if personnel else "",
        "workflow_type": a.workflow_type,
        "submitter_id": a.submitter_id,
        "submitter_name": submitter.real_name if submitter else "",
        "approver_id": a.approver_id,
        "approver_name": approver.real_name if approver else "",
        "status": a.status,
        "change_data": a.change_data,
        "submit_time": a.submit_time,
        "approve_time": a.approve_time,
        "approve_comment": a.approve_comment,
        "withdrawn_time": a.withdrawn_time,
        "created_at": a.created_at,
    }


def approve(db: Session, approval_id: int, comment: str) -> WfApproval | None:
    """Approve: apply changes to personnel, set status, return updated approval."""
    from app.services.push_service import execute_push

    a = db.get(WfApproval, approval_id)
    if not a or a.status != "pending":
        return None

    a.status = "approved"
    a.approve_time = datetime.now(timezone.utc)
    a.approve_comment = comment

    # Apply the change data to personnel
    personnel = db.get(MdmPersonnel, a.personnel_id)
    if personnel and a.change_data:
        personnel_service.apply_changes(db, personnel, a.change_data)

    db.commit()
    db.refresh(a)

    # Trigger simulated push to downstream systems
    execute_push(db, a)

    return a


def reject(db: Session, approval_id: int, comment: str) -> WfApproval | None:
    """Reject: discard changes, reset personnel status, return updated approval."""
    a = db.get(WfApproval, approval_id)
    if not a or a.status != "pending":
        return None

    a.status = "rejected"
    a.approve_time = datetime.now(timezone.utc)
    a.approve_comment = comment

    # If this was an update (not create), revert personnel status
    personnel = db.get(MdmPersonnel, a.personnel_id)
    if personnel:
        if a.workflow_type == "create":
            # For rejected creates, mark the personnel as inactive
            personnel.status = "inactive"
        else:
            personnel.status = "active"

    db.commit()
    db.refresh(a)
    return a


def withdraw(db: Session, approval_id: int, user_id: int) -> WfApproval | None:
    """Withdraw: only allowed if current user is submitter and status is pending."""
    a = db.get(WfApproval, approval_id)
    if not a or a.status != "pending":
        return None
    if a.submitter_id != user_id:
        return None

    a.status = "withdrawn"
    a.withdrawn_time = datetime.now(timezone.utc)

    # Revert personnel status
    personnel = db.get(MdmPersonnel, a.personnel_id)
    if personnel:
        if a.workflow_type == "create":
            personnel.status = "inactive"
        else:
            personnel.status = "active"

    db.commit()
    db.refresh(a)
    return a


def _find_approver(db: Session) -> int | None:
    """Find the first active approver user. In a real system this would use configurable rules."""
    from sqlalchemy import select as s
    approver = db.execute(
        s(SysUser).where(SysUser.role == "approver", SysUser.status == "active").limit(1)
    ).scalar()
    return approver.id if approver else None
